jobs/e2e-test.groovy

```diff
@@ -302,6 +305,19 @@ def runTestServer() {
       // before it does so, so it doesn't read "old" data.
       sh("rm -rf genfiles/test-reports");
 
+      tests = sh(script: "testing/runtests_server.py -n ${exec.shellEscapeList(runSmokeTestsArgs)} . ", returnStdout: true);
+      // The runtests_server.py script with -n outputs all tests to run of 
+      // various types. We only need to worry about smoke tests, which are 
+      // also the last section, so grab everything after the header. If it's 
+      // empty, there are no tests to run.
```

Comment:
This is fragile! -- the output of `-n` isn't meant to be machine-readable.  I'd at least put a pointer in runtests_server.py pointing to this, if someone wants to change the output of the script.  And maybe make sure runtests_server promises that the smoketests are last.

More robust would be to do a something like this:
   testing/runtests_server.py -n <args> . | sed -n /^SMOKE TESTS:/,/:/p | grep -v $ | xargs

This will get everything between the "SMOKE TESTS" section and next section (or end of output), and trim whitespace.  (You could also keep the trimming in groovy, of course.) 
 Then the isAllWhitespace() check should work really reliably; you'd just need to comment in runtests_server.py not to change the labels of the test types.

Reply:
> This is fragile! -- the output of `-n` isn't meant to be machine-readable. I'd at least put a pointer in runtests_server.py pointing to this, if someone wants to change the output of the script. And maybe make sure runtests_server promises that the smoketests are last.

I agree! A json output would certainly be more usable! I can write up a ticket to capture that work, but it's not in this repo so would really be better asynchronously done to this work.

> More robust would be to do a something like this:
> testing/runtests_server.py -n . | sed -n /^SMOKE TESTS:/,/:/p | grep -v $ | xargs
> 
> This will get everything between the "SMOKE TESTS" section and next section (or end of output), and trim whitespace. (You could also keep the trimming in groovy, of course.)
> Then the isAllWhitespace() check should work really reliably; you'd just need to comment in runtests_server.py not to change the labels of the test types.

I think we could get to that same level of robustness just by adding a further substring method that trims off everything after an empty newline (if there is one), and doing it in the groovy would probably be more readable than sed (at least to me, I suppose I can't speak to other maintainers). Also it appears there's a problem using pipes in the shell commands, I tried a similar solution to what you proposed here first and it consistently failed the whole job.

Reply:
@dbraley :+1: to string manipulation in a programming language being more readable (and discoverable, even) than shelling out to sed.

Reply:
I would say that's true for most programming languages, but not for groovy. :-/  That said, it's up to you.

> I think we could get to that same level of robustness just by adding a further substring method that trims off everything after an empty newline (if there is one)

That works too!